openwisp_firmware_upgrader/upgraders/openwrt.py

```diff
@@ -155,8 +155,38 @@ def upload(self, image_file, remote_path):
         except Exception as e:
             raise RecoverableFailure(str(e))
 
+    def _verify_device_uuid(self):
+        """
+        Verifies that the UUID of the device being upgraded matches
+        the UUID in the device's configuration
+        """
+        device_uuid = self.upgrade_operation.device.pk
+        # Get UUID from device's openwisp config
+        output, exit_code = self.exec_command(
+            'uci get openwisp.http.uuid', exit_codes=[0, 1]
+        )
+        if exit_code == 1 or not output.strip():
+            self.log(_('Could not read device UUID from configuration'))
```

Comment:
I haven't seen any assertion verifying for the presence of this log line, did you add any?.